fiaas_skipper/deploy/deploy.py

```diff
@@ -155,3 +137,43 @@ def _get_status(dep, app):
     except TypeError:
         return Status('UNAVAILABLE', 'Unable to determine available/ready replicas from k8s server')
     return Status('OK', '')
+
+
+class StatusTracker(Thread):
+    def __init__(self, cluster, application):
+        Thread.__init__(self)
+        self._cluster = cluster
+        self._status = {}
+        self._application = application
+        self._statuslock = Lock()
+
+    def __call__(self):
+        with self._statuslock:
+            return tuple(self._status.values())
+
+    def _get_status(self, namespace=None):
+        deployment_configs = self._cluster.find_deployment_configs(NAME, namespace=namespace)
+        deployments = {d.metadata.namespace: d for d in Deployment.find(NAME, namespace=namespace)}
+        applications = {d.metadata.namespace: d for d in self._application.find(NAME, namespace=namespace)}
+        res = []
+        for c in deployment_configs:
+            dep = deployments.get(c.namespace)
+            app = applications.get(c.namespace)
+            version = _get_version(dep)
+            status = _get_status(dep, app)
+            res.append(DeploymentStatus(name=NAME,
+                                        namespace=c.namespace,
+                                        status=status.summary,
+                                        description=status.description,
+                                        version=version or '',
+                                        channel=c.tag or ''))
+        return res
+
+    def _update_status(self):
+        with self._statuslock:
+            self._status = {status.namespace: status for status in self._get_status()}
```

Comment:
This holds the lock for the entire time it spends getting data from the API, essentially blocking the frontend from getting the current list of statuses until it is done. I think it should get the data, then grab the lock before updating the `self._status` dictionary. So something like this:

``` 
def _update_status(self):
    new_status = self._get_status()
    with self._statuslock:
        self._status = {status.namespace: status for status in new_status}
```

Reply:
Done